### DISCUSSION TITLE
How can I make this tr command work in Xonsh?

### DISCUSSION THREAD

**Discussion by ksaj** [xonsh/xonsh#5234] [Q&A] [answered]
In BASH:
> head /dev/urandom | tr -dc [:digit:]
72438490526699619204315035199725308989775086689836882186584692324688968729817859295703510531533245501466087301693176717747067445139636359217

In Xonsh, attempt 1:
> head /dev/urandom |tr -dc [:digit:]
NameError: name 'head' is not defined

In Xonsh, attempt 2:
> /usr/bin/head /dev/urandom |tr -dc [:digit:]
........................
  File "<stdin>", line 1
    /usr/bin/head /dev/urandom |tr -dc [:digit:]
SyntaxError: ('code: /',)

In Xonsh, attempt 3:
> /usr/bin/head /dev/urandom |/usr/bin/tr -dc [:digit:]
........................
  File "<stdin>", line 1
    /usr/bin/head /dev/urandom |/usr/bin/tr -dc [:digit:]
SyntaxError: ('code: /',)

Just about everything else I do gets the similar response as Attempt 3.  Pipes seem to work otherwise, but this is beating me up!

EDIT: Sorry for all the edits. Was having pasting issues. Ignore previous versions of this post.

**Comment by Qyriad** [comment, marked as answer]
Your first one works if you put quotes around `[:digit:]`:

```sh
head /dev/urandom | tr -dc "[:digit:]"
```

**Reply by ksaj** [reply]
Ah, that's the ticket. Thanks! I was thinking the quotes would stop it from being interpreted by tr. I think I was already rattled by the 'head' not working as-is since it is also a python keyword, so I didn't get around to quotes. I'm guessing it thinks the colon is part of python code.

Is it a bug, a feature, or a side effect?

**Reply by anki-code** [reply]
Hello @ksaj ! Please start here - https://github.com/anki-code/xonsh-cheatsheet/blob/main/README.md#three-most-frequent-things-that-newcomers-overlook

**Reply by Qyriad** [reply]
It has nothing to do with `head`, or even the colon. It's just the brackets. Try `echo [ 1 2 3 ]` and you'll see what I mean.

Though I'm not really sure *why*, since I don't think brackets do anything in subprocess mode without a preceding `!` or `$`? @anki-code is there a reason for brackets needing quotes that I'm missing?

**Reply by ksaj** [reply]
This is what throws me for such a loop:

@ echo [1, 2, 3, 4]
.............................................
NameError: name 'echo' is not defined
@ echo "hi"
hi

I get that it wants quotes, but this is weirdness.

**Reply by anki-code** [reply]
I also don't know why :) I assume that xonsh try to be as predictable as possible as well as reduce future issues i.e. any complex argument should be quoted.
